zkevm-circuits/src/evm_circuit/util.rs

```diff
@@ -499,10 +499,179 @@ impl<F: FieldExt, const N: usize> Expr<F> for RandomLinearCombination<F, N> {
     }
 }
 
-pub(crate) type Word<F> = RandomLinearCombination<F, 32>;
+#[derive(Clone, Debug)]
+pub(crate) struct WordLimbs<T, const N: usize> {
+    pub limbs: [T; N],
+}
+
+pub(crate) type Word2<T> = WordLimbs<T, 2>;
+
+pub(crate) type Word4<T> = WordLimbs<T, 4>;
+
+pub(crate) type Word16<T> = WordLimbs<T, 16>;
+
+pub(crate) type Word32<T> = WordLimbs<T, 32>;
+
+impl<T, const N: usize> WordLimbs<T, N> {
+    fn new(limbs: [T; N]) -> Self {
+        Self { limbs }
+    }
+
+    fn n() -> usize {
+        N
+    }
+}
+
+impl<T: Default, const N: usize> Default for WordLimbs<T, N> {
+    fn default() -> Self {
+        Self {
+            limbs: [(); N].map(|_| T::default()),
+        }
+    }
+}
+
+pub(crate) trait WordCells<F: FieldExt, const N: usize> {
+    fn assign<const N1: usize>(
+        &self,
+        region: &mut CachedRegion<'_, '_, F>,
+        offset: usize,
+        bytes: Option<[u8; N1]>,
+    ) -> Result<Vec<AssignedCell<F, F>>, Error>;
+
+    fn expr(&self) -> WordLimbs<Expression<F>, N>;
+}
+
+impl<F: FieldExt, const N: usize> WordCells<F, N> for WordLimbs<Cell<F>, N> {
+    fn assign<const N1: usize>(
+        &self,
+        region: &mut CachedRegion<'_, '_, F>,
+        offset: usize,
+        bytes: Option<[u8; N1]>,
+    ) -> Result<Vec<AssignedCell<F, F>>, Error> {
+        assert_eq!(N1 % N, 0); // TODO assure N|N1, find way to use static_assertion instead
+        bytes.map_or(Err(Error::Synthesis), |bytes| {
+            bytes
+                .chunks(N1 / N) // chunk in little endian
+                .map(|chunk| from_bytes::value(chunk))
+                .zip(self.limbs.iter())
+                .map(|(value, cell)| cell.assign(region, offset, Value::known(value)))
+                .collect()
+        })
+    }
+
+    fn expr(&self) -> WordLimbs<Expression<F>, N> {
+        return WordLimbs::new(self.limbs.map(|cell| cell.expr()));
+    }
+}
+
+// `Word`, special alias for Word2.
```

Comment:
I'm a little bit confused, I thought the number `2` stands for the number of byte. So, `Word2` is a 2 bytes data structure, isn't it? And from my understanding, word hi/lo is to represent 32 bytes evm word. Could you please explain it more? thanks

Reply:
I moved the new `Word` type to new file `word.rs`, and add simple comment.

Word`N` means splits 256 bit into N chunk. Each chunk is also called limb. (would chunk be better naming 😢 ?)

Therefore, Word2 means 2 limb, each with 128bit

Reply:
haha, got it. I thought the other way around. I thought `Word32` includes 32 limbs and each limb is 1 byte. So, `Word2` is a 2 bytes data.

Reply:
What if we name these types like this,
- `Word` and `Word128` stands for 2 x 128 bits limbs
- `Word64` stands for 4 x `64` bits limbs
- `Word16` stands for 16 x `16` bits limbs
- `Word8`   stands for 32 x `8` bits limbs

Just like `uintN` in solidity, the following `N` stands for bit numbers

Reply:
I want to say that I like the original nomenclature that @hero78119 proposed (where `N` is the number of limbs).
If `N` means number of bits, I find it easier to mistake it for the word size, like the `uintN` example; but that's not the case: all Words are 32 bytes in Ethereum.
Also I think it's more direct to see `WordN` and then inspect the type and see an array of length `N`.
The other way would be inspecting the type and seeing an array of length `256 / N`